api/agent/drivers/docker/docker_client.go

```diff
@@ -187,15 +252,34 @@ func (d *dockerWrap) Info(ctx context.Context) (info *docker.DockerInfo, err err
 	return d.docker.Info()
 }
 
-func (d *dockerWrap) AttachToContainerNonBlocking(ctx context.Context, opts docker.AttachToContainerOptions) (docker.CloseWaiter, error) {
-	ctx, span := trace.StartSpan(ctx, "docker_attach_container")
-	defer span.End()
-	return d.docker.AttachToContainerNonBlocking(opts)
+func (d *dockerWrap) AttachToContainer(ctx context.Context, opts docker.AttachToContainerOptions) (docker.CloseWaiter, error) {
+	ctx, closer := makeTracker(ctx, "docker_attach_container")
+	defer closer()
+
+	if opts.Success != nil {
+		logrus.Fatal("BUG: Invalid AttachToContainerOptions, Success channel must not be set")
+	}
+	opts.Success = make(chan struct{})
+
+	// We use non-blocking here since we need the CloseWaiter
+	waiter, err := d.docker.AttachToContainerNonBlocking(opts)
+
+	// Sync up with NB Attacher above before starting the task
+	if err == nil {
+		// WARNING: the I/O below requires docker hijack function to honor
+		// the contract below, specifically if an error is not returned
+		// from AttachToContainerNonBlocking, then max blocking time
+		// here should be what drv.docker dialer/client config was set to.
```

Comment:
i am not sure we are setting this anymore either, not sure this assumption is safe. i can double check through this tomorrow hopefully

Reply:
So, I got this queued up: https://github.com/fnproject/fn/pull/1201
I'm not sure if it's a significant improvement, but at least it honors the context. I think we used to set some timeouts in the dialers in the driver, but now, I think hijack & Dialer/D combo may indeed block without. If that happens, then the dialer go-routine would block/leak, along with our go-routine in #1201. What do you think? The rationale here (AFAIR) was that we couldn't start the task before these streams were attached. (Might be less of a problem with docker-syslog and upcoming socket-IO, but it probably safer to leave it in.)